File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
type Color = (typeof COLORS)[number];

function format(value: number, unit: string): string {
  return `${value} ${unit}`;
}

export function decodedResistorValue(colors: Color[]) {
  const [c1, c2, c3] = colors;
  const zeros = COLORS.indexOf(c3);

  let value =
    COLORS.indexOf(c1) * 10 ** (zeros + 1) +
    COLORS.indexOf(c2) * (zeros > 0 ? 10 ** zeros : 1);

  if (value < 1_000) return format(value, "ohms");

  if (value >= 1_000 && value < 10 ** 6) {
    return format(value / 1_000, "kiloohms");
  }

  if (value >= 10 ** 6 && value < 10 ** 9) {
    return format(value / 10 ** 6, "megaohms");
  }

  return format(value / 10 ** 9, "gigaohms");
}

export const COLORS = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white",
] as const;

File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @codes %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label([a, b, n | _]) do
    value = (@codes[a] * 10 + @codes[b]) * 10 ** @codes[n]

    case value do
      x when x >= 1_000 -> {x / 1_000, :kiloohms}
      x -> {x, :ohms}
    end
  end
end
